cdb2api/cdb2api.c

```diff
@@ -5056,6 +5056,87 @@ int cdb2_bind_index(cdb2_hndl_tp *hndl, int index, int type,
+                    int type, const void *varaddr[], int typelen, int length[])
+{
+    hndl->n_bindvars++;
+    hndl->bindvars = realloc(hndl->bindvars, sizeof(CDB2SQLQUERY__Bindvalue *) *
+                                                 hndl->n_bindvars);
+    CDB2SQLQUERY__Bindvalue *bindval = malloc(sizeof(CDB2SQLQUERY__Bindvalue));
+    cdb2__sqlquery__bindvalue__init(bindval);
+    bindval->type = type;
+    bindval->varname = (char *)varname;
+    switch(type) {
+    case CDB2_INTEGER:
+        if (typelen == sizeof(int32_t)) {
+            bindval->int64_arr = malloc(sizeof(CDB2SQLQUERY__Bindvalue__Int64array));
+            cdb2__sqlquery__bindvalue__int64array__init(bindval->int64_arr);
+            bindval->int64_arr->elements = malloc(count * sizeof(int64_t));
+            for(int i=0; i<count;i++) {
+                bindval->int64_arr->elements[i]= ((int32_t*)varaddr)[i];
+            }
+            bindval->int64_arr->n_elements = count;
+        }
+        else if (typelen == sizeof(int64_t)) {
+            bindval->int64_arr = malloc(sizeof(CDB2SQLQUERY__Bindvalue__Int64array));
+            cdb2__sqlquery__bindvalue__int64array__init(bindval->int64_arr);
+            bindval->int64_arr->elements = malloc(count * sizeof(int64_t));
+            memcpy(bindval->int64_arr->elements, varaddr, count * typelen);
+            bindval->int64_arr->n_elements = count;
+        }
+    break;
+    case CDB2_REAL:
+        if (typelen == sizeof(float)) {
+            bindval->real64_arr = malloc(sizeof(CDB2SQLQUERY__Bindvalue__Real64array));
+            cdb2__sqlquery__bindvalue__real64array__init(bindval->real64_arr);
+            bindval->real64_arr->elements = malloc(count * typelen);
+            //printf("AZ: ");
+            for(int i=0; i<count;i++) {
+                bindval->real64_arr->elements[i] = ((float*)varaddr)[i];
+                //printf("%ld ", bindval->real64_arr->elements[i]);
+            }
+            //printf("count %d\n", count);
+            bindval->real64_arr->n_elements = count;
+        }
+        else if (typelen == sizeof(double)) {
+            bindval->real64_arr = malloc(sizeof(CDB2SQLQUERY__Bindvalue__Real64array));
+            cdb2__sqlquery__bindvalue__real64array__init(bindval->real64_arr);
+            bindval->real64_arr->elements = malloc(count * typelen);
+            memcpy(bindval->real64_arr->elements, varaddr, count * typelen);
+            bindval->real64_arr->n_elements = count;
+        }
+    break;
+    case CDB2_CSTRING:
+        bindval->text_arr = malloc(sizeof(CDB2SQLQUERY__Bindvalue__Textarray));
+        cdb2__sqlquery__bindvalue__textarray__init(bindval->text_arr);
+        bindval->text_arr->elements = malloc(count * sizeof(char *));
+        //printf("AZ: ");
+        for(int i=0; i<count;i++) {
+            bindval->text_arr->elements[i] = strdup(((char**)varaddr)[i]); // TODO: check that this gets freed
```

Comment:
One suggestion that was made internally was to get the length from the `lengths` argument if non-null for strings.

Reply:
Sure, sounds fine.

Reply:
Actually, that makes it behave like a blob, and after further discussion, we would not support blobs as native type. If strings are not null terminated, they won't pack correctly by protobuf.